[inductor] Add a subprocess-based parallel compile

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126817
* __->__ #126816

Summary:
Adds a "safe" parallel compile implementation that a) Popens a sub-process with an entry point we control, and b) Uses a ProcessPoolExecutor in that sub-processes to perform parallel compiles. This change essentially squashes these two implementations from jansel, but removes the "thread-based" approach since benchmarking revealed that compile-time performance was poor compared to the existing impl:
https://github.com/pytorch/pytorch/pull/124682
https://github.com/pytorch/pytorch/pull/122941

This PR adds the implementation, but defaults to the existing "fork". I'll submit a separate change to enable.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

masnesral committed:
[inductor] Add a subprocess-based parallel compile

Summary:
Adds a "safe" parallel compile implementation that a) Popens a sub-process with an entry point we control, and b) Uses a ProcessPoolExecutor in that sub-processes to perform parallel compiles. This change essentially squashes these two implementations from jansel, but removes the "thread-based" approach since benchmarking revealed that compile-time performance was poor compared to the existing impl:
https://github.com/pytorch/pytorch/pull/124682
https://github.com/pytorch/pytorch/pull/122941

This PR adds the implementation, but defaults to the existing "fork". I'll submit a separate change to enable.

[ghstack-poisoned]

masnesral commented:
Benchmark results: https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C+14+May+2024+19%3A50%3A25+GMT&stopTime=Tue%2C+21+May+2024+19%3A50%3A25+GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh%2Fmasnesral%2F61%2Fhead&lCommit=e5e2e5796b0ffab0922da6a7a714a26ae57ba513&rBranch=main&rCommit=53f73cdeb6fe53155f91064ca15b722e80dec2f3

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -3177,6 +3137,8 @@ def wait(self, scope: Dict[str, Any]) -> None:
 if (
     os.environ.get("TORCH_TNT_IN_USE", "0") == "1"
     or os.environ.get("TORCH_WARM_POOL", "1") != "1"
+    # TODO(masnesral): warm pool is broken, need to fix it
```
@jansel, I commandeered this TODO you had originally, but I'm not sure there's anything to actually do here? We're calling _warm_process_pool() on the ProcessPoolExecutor owned by the subproc in SubprocMain.__init__(). Or did you have some other kind of warmup?

jansel replied:
I was getting errors with this turned on.  Does it pass for you?

masnesral committed:
Update on "[inductor] Add a subprocess-based parallel compile"


Summary:
Adds a "safe" parallel compile implementation that a) Popens a sub-process with an entry point we control, and b) Uses a ProcessPoolExecutor in that sub-processes to perform parallel compiles. This change essentially squashes these two implementations from jansel, but removes the "thread-based" approach since benchmarking revealed that compile-time performance was poor compared to the existing impl:
https://github.com/pytorch/pytorch/pull/124682
https://github.com/pytorch/pytorch/pull/122941

This PR adds the implementation, but defaults to the existing "fork". I'll submit a separate change to enable.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/compile_worker/__main__.py:
```diff
@@ -0,0 +1,45 @@
+import sys
+import typing
+
+from torch._inductor.codecache import _set_triton_ptxas_path, caching_device_properties
```
`_set_triton_ptxas_path` isn't defined in torch._inductor.codecache, do you mean the one defined in compile_tasks.py?

masnesral replied:
Hmmm; I guess it works because codecache.py imports that function. But yeah, will fix

jamesjwu replied:
python torch/_inductor/compile_worker/__main__.py doesn't seem to run for me if you just call it directly, not sure if there's some special way to call it so it passes today

masnesral replied:
I'm perplexed. Seems to work fine for me:
```
$ python torch/_inductor/compile_worker/__main__.py --help
usage: __main__.py [-h] [--workers WORKERS] [--parent PARENT]

options:
  -h, --help         show this help message and exit
  --workers WORKERS
  --parent PARENT
```

(but making change anyway)

eellison commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -0,0 +1,254 @@
+    # We have to fork processes for compiler workers, but the more memory and other resources that are loaded, the
+    # slower the os.fork time is, quite drastically. It also holds the GIL so we can't put it on another thread.
+
+    # Examples:
+    # A simple x + x + x script: 10ms seconds in the middle of the program, 2ms at startup
+    # tf_efficientnet_b0 benchmark: 50ms! in the middle of the program , 3ms at startup
+
+    # So we want to start the workers early when it is still cheap, and also to allow the workers to get
+    # ready before we have work for them.
+
+    # ProcessPoolExecutor also does not launch the workers until it finds a point when all the workers are idle.
+    # But if we waited until then fork time will be long and we will be waiting for the processes to initialize.
```
Are these comments still relevant ?

masnesral replied:
Lemme add a TODO and I'll benchmark

masnesral committed:
Update on "[inductor] Add a subprocess-based parallel compile"


Summary:
Adds a "safe" parallel compile implementation that a) Popens a sub-process with an entry point we control, and b) Uses a ProcessPoolExecutor in that sub-processes to perform parallel compiles. This change essentially squashes these two implementations from jansel, but removes the "thread-based" approach since benchmarking revealed that compile-time performance was poor compared to the existing impl:
https://github.com/pytorch/pytorch/pull/124682
https://github.com/pytorch/pytorch/pull/122941

This PR adds the implementation, but defaults to the existing "fork". I'll submit a separate change to enable.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
New benchmark result after properly warming the cache:

https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2021%20May%202024%2003%3A36%3A49%20GMT&stopTime=Tue%2C%2028%20May%202024%2003%3A36%3A49%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/masnesral/65/head&lCommit=4293f564e1bbd0724578cecc05e6159b4166ce93&rBranch=main&rCommit=b516de8cac9317155104d476056bae1d910a125c

masnesral commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert -m "causing internal regression" -c "ghfirst"
